ent/ent.go

```diff
@@ -0,0 +1,92 @@
+//
+// Copyright 2022 SkyAPM org
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package ent
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/SkyAPM/go2sky"
+	"github.com/SkyAPM/go2sky-plugins/ent/gen/entschema"
+
+	"entgo.io/ent/dialect"
+	entsql "entgo.io/ent/dialect/sql"
+	sqlPlugin "github.com/SkyAPM/go2sky-plugins/sql"
+)
+
+// Driver custom wrapper ent driver, Don't use native.
+type Driver struct {
+	dialect.Driver
+	db *sqlPlugin.DB
+}
+
+// NewDriver EntDriver.
+func NewDriver(drv dialect.Driver, db *sqlPlugin.DB) *Driver {
+	return &Driver{drv, db}
+}
+
+// BeginTx custom.
+func (e *Driver) BeginTx(ctx context.Context, option *sql.TxOptions) (dialect.Tx, error) {
+	tx, err := e.db.BeginTx(ctx, option)
+	if err != nil {
+		return nil, err
+	}
+	return &entsql.Tx{
+		Conn: entsql.Conn{ExecQuerier: tx},
+		Tx:   tx,
+	}, nil
+}
+
+// NewEntClient ent client.
+func NewEntClient(dsn string, tracer *go2sky.Tracer) *entschema.Client {
+	driver, url := "mysql", dsn
+	/*
+		db, err := sql.Open(driver, url)
+		if err != nil {
+			panic(err)
+		}
+		drv := entsql.OpenDB(driver, db)
+	*/
+	drv, err := entsql.Open(driver, url)
+	if err != nil {
+		panic(err)
+	}
+
+	apmDB, err := sqlPlugin.Open(
+		driver,
+		url,
+		tracer,
+		sqlPlugin.WithSQLDBType(sqlPlugin.MYSQL),
+		sqlPlugin.WithQueryReport(),
+		sqlPlugin.WithParamReport(),
+	)
+	if err != nil {
+		panic(err)
+	}
+	apmDB.DB.SetMaxIdleConns(50)
+	apmDB.DB.SetMaxOpenConns(200)
+
+	drv.Conn = entsql.Conn{ExecQuerier: apmDB}
+
+	EntClient := entschema.NewClient(
+		entschema.Driver(
+			NewDriver(drv, apmDB),
+		),
+	)
```

Comment:
@xiaohuo473225193 The `gen/entschema` seems to be generated based on [schema](https://github.com/SkyAPM/go2sky-plugins/pull/52/files#diff-c5fc389a809e819c5217697a1711b202bfb7f5594a14da80cdabad8b8c6a939cR28-R30),  this plugin does not apply to other schema, right？

Reply:
No. The plugin is used for all schema definitions.

Reply:
But, The [NewClient](https://github.com/SkyAPM/go2sky-plugins/pull/52/files#diff-22e40780487774f76e02ccbc532177f7480a6b484b1dcfddb1fe2d5ed4145aa3R42-R53) that you have here can only be used for the [user schema](https://github.com/SkyAPM/go2sky-plugins/pull/52/files#diff-c5fc389a809e819c5217697a1711b202bfb7f5594a14da80cdabad8b8c6a939c) that you define.

Reply:
Ent can only operate based on the code it generates. NewClient it's the only way in method.

Reply:
`NewClient` has nothing to do with the schema definition.

Reply:
> Ent can only operate based on the code it generates. NewClient it's the only way in method.

When I create a new schema, I need to regenerate the code, right?

Reply:
Yes. But schema change affects only the schema model. The Ent code itself can be considered fixed.

Reply:
Can you demonstrate how to reference ent plugin in the new repository?

Reply:
This is actually an example, not a specific plugin. This is just one way to access the trace method.

Reply:
It is just a reference method provided by Ent access trace. I use this approach in my project.

Reply:
why stop？